

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>velocyto.counter &mdash; velocyto 0.9.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="velocyto 0.9.5 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> velocyto
          

          
          </a>

          
            
            
              <div class="version">
                0.9
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome to velocyto.py!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../install/index.html">Installation Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial/index.html">Tutorial</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../fullapi/index.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks/index.html">Notebooks</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../citing/index.html">Authors and citations</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../license/index.html">License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">velocyto</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>velocyto.counter</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for velocyto.counter</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="k">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="kn">import</span> <span class="nn">velocyto</span> <span class="k">as</span> <span class="nn">vcy</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">tagpat</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s2">&quot;NH:i:1.+CB:Z:([ACGT]+).+UB:Z:([ACGT]+)&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="ExInCounter"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter">[docs]</a><span class="k">class</span> <span class="nc">ExInCounter</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot; Main class to do the counting of introns and exons &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">valid_bcs2idx</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; valid_bcs2idx is a dict mapping valid cell barcodes to unique arbitrary numeric indexes used in the output arrays &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">valid_bcs2idx</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filter_mode</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span> <span class="o">=</span> <span class="n">valid_bcs2idx</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filter_mode</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>  <span class="c1"># type: Dict[str, List]</span>
        <span class="c1"># NOTE: by using a default dict and not logging access to keys that do not exist, we might miss bugs!!!</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># type: List</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>  <span class="c1"># type: Dict[str, List]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">bclen</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
        <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

<div class="viewcode-block" id="ExInCounter.process"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.process">[docs]</a>    <span class="k">def</span> <span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ivlfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; Performs all steps in the right order &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">read_genes</span><span class="p">(</span><span class="n">ivlfile</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mark_up_introns</span><span class="p">(</span><span class="n">samfile</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">samfile</span><span class="p">)</span></div>

<div class="viewcode-block" id="ExInCounter.parse_cigar_tuple"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.parse_cigar_tuple">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">parse_cigar_tuple</span><span class="p">(</span><span class="n">cigartuples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">],</span> <span class="n">pos</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
        <span class="n">segments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">spliced</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">clip5</span> <span class="o">=</span> <span class="n">clip3</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="k">for</span> <span class="n">operation_id</span><span class="p">,</span> <span class="n">length</span> <span class="ow">in</span> <span class="n">cigartuples</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># vcy.CIGAR[operation_id] == &quot;BAM_CMATCH&quot;</span>
                <span class="n">segments</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">p</span><span class="p">,</span> <span class="n">p</span> <span class="o">+</span> <span class="n">length</span><span class="p">))</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>  <span class="c1"># A splice || vcy.CIGAR[operation_id] == &#39;BAM_CREF_SKIP&#39;</span>
                <span class="n">spliced</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>  <span class="c1"># A deletion || cy.CIGAR[operation_id] == &#39;BAM_CDEL&#39;</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># bases at 5&#39; or 3&#39; are NOT part of the alignment || vcy.CIGAR[operation_id] == &#39;BAM_CSOFT_CLIP&#39;</span>
                <span class="k">if</span> <span class="n">p</span> <span class="o">==</span> <span class="n">pos</span><span class="p">:</span>
                    <span class="n">clip5</span> <span class="o">=</span> <span class="n">length</span>  <span class="c1"># At start of alignment</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">clip3</span> <span class="o">=</span> <span class="n">length</span>  <span class="c1"># Must be at end of alignment vcy.CIGAR[operation_id] in [&quot;BAM_CINS&quot;, &quot;BAM_CHARD_CLIP&quot;]</span>
                <span class="n">p</span> <span class="o">+=</span> <span class="n">length</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># An insertion BAM_CINS</span>
                <span class="k">pass</span>  <span class="c1"># maybe there should be a limit lenght accepted?</span>
            <span class="k">elif</span> <span class="n">operation_id</span> <span class="o">==</span> <span class="mi">5</span><span class="p">:</span>  <span class="c1"># BAM_CHARD_CLIP</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Hard clip was encountered! All mapping are assumed soft clipped&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">segments</span><span class="p">,</span> <span class="n">spliced</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span></div>

<div class="viewcode-block" id="ExInCounter.iter_unique_alignments"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.iter_unique_alignments">[docs]</a>    <span class="k">def</span> <span class="nf">iter_unique_alignments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">yield_line</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Reading </span><span class="si">{samfile}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">fin</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">samfile</span><span class="p">)</span>  <span class="c1"># type: pysam.AlignmentFile</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">read</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fin</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_unmapped</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">bc</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s2">&quot;CB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;-1&quot;</span><span class="p">)</span>
                <span class="n">umi</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s2">&quot;UB&quot;</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">bc</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_mode</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span><span class="p">[</span><span class="n">bc</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">counter</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="s1">&#39;-&#39;</span> <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">is_reverse</span> <span class="k">else</span> <span class="s1">&#39;+&#39;</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="n">fin</span><span class="o">.</span><span class="n">get_reference_name</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">rname</span><span class="p">)</span>  <span class="c1"># this is return a string otherwise read.rname for the integer</span>
            <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">):</span>
                <span class="c1"># I have to deal with incongruences with the .gft (what is cellranger doing???)</span>
                <span class="c1"># NOTE Why this happens?</span>
                <span class="k">if</span> <span class="s2">&quot;_&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
                    <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;M&quot;</span><span class="p">:</span>
                        <span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;MT&quot;</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">pos</span>
            <span class="n">segments</span><span class="p">,</span> <span class="n">spliced</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parse_cigar_tuple</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">cigartuples</span><span class="p">,</span> <span class="n">pos</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">segments</span> <span class="o">==</span> <span class="p">[]:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;No segments in read:</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">10000000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Read first {i // 1000000} million mapped reads&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">yield_line</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">(</span><span class="n">bc</span><span class="p">,</span> <span class="n">umi</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">cigarstring</span><span class="p">,</span> <span class="n">segments</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span><span class="p">,</span> <span class="n">spliced</span><span class="p">),</span> <span class="n">read</span><span class="o">.</span><span class="n">tostring</span><span class="p">(</span><span class="n">fin</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Read</span><span class="p">(</span><span class="n">bc</span><span class="p">,</span> <span class="n">umi</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">cigarstring</span><span class="p">,</span> <span class="n">segments</span><span class="p">,</span> <span class="n">clip5</span><span class="p">,</span> <span class="n">clip3</span><span class="p">,</span> <span class="n">spliced</span><span class="p">)</span>
        <span class="n">fin</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="ExInCounter.read_repeats"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.read_repeats">[docs]</a>    <span class="k">def</span> <span class="nf">read_repeats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; Read masked repeats &quot;&quot;&quot;</span>
        <span class="n">fin</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">gtf_file</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fin</span><span class="p">:</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">fields</span>
            <span class="c1"># Properly format the chromosome to match how it is parsed in the bam file</span>
            <span class="k">if</span> <span class="s2">&quot;_&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">:</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
                <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;M&quot;</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;MT&quot;</span>
            <span class="n">ivlstart</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
            <span class="n">ivlend</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
            <span class="n">chromstrand</span> <span class="o">=</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span>
            <span class="n">ivl</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Interval</span><span class="p">(</span><span class="n">ivlstart</span><span class="p">,</span> <span class="n">ivlend</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;mm10_rmsk&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="p">[</span><span class="n">chromstrand</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ivl</span><span class="p">)</span>

        <span class="c1"># Make sure they are sorted (they should be becouse of the way the files where genrated)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chromstrand</span><span class="p">,</span> <span class="n">ivls</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">ivls</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>  <span class="c1"># relies on the __lt__ method of vcy.Interval</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ivls</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">n</span></div>

<div class="viewcode-block" id="ExInCounter.read_genes"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.read_genes">[docs]</a>    <span class="k">def</span> <span class="nf">read_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ivlfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; Read genes and intervals &quot;&quot;&quot;</span>
        <span class="n">nbcs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">ivlfile</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">trname</span><span class="p">,</span> <span class="n">trid</span><span class="p">,</span> <span class="n">genename</span><span class="p">,</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">strand</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[:</span><span class="mi">6</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">):</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
            <span class="n">chromstrand</span> <span class="o">=</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">strand</span>
            <span class="n">ivls</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># type: List</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Gene</span><span class="p">(</span><span class="n">genename</span><span class="p">,</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">nbcs</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">genestart</span><span class="p">,</span> <span class="n">geneend</span> <span class="o">=</span> <span class="mi">2</span><span class="o">**</span><span class="mi">30</span><span class="p">,</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)):</span>
                <span class="n">field</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">span</span><span class="p">,</span> <span class="n">ivltype</span><span class="p">,</span> <span class="n">other_exons</span><span class="p">,</span> <span class="n">other_introns</span> <span class="o">=</span> <span class="n">field</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)</span>
                <span class="n">ivlstart</span><span class="p">,</span> <span class="n">ivlend</span> <span class="o">=</span> <span class="n">span</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
                <span class="n">ivlstart</span><span class="p">,</span> <span class="n">ivlend</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">ivlstart</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ivlend</span><span class="p">)</span>
                <span class="n">genestart</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">genestart</span><span class="p">,</span> <span class="n">ivlstart</span><span class="p">)</span>
                <span class="n">geneend</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">geneend</span><span class="p">,</span> <span class="n">ivlend</span><span class="p">)</span>
                <span class="n">is_last_3exon_flag</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">i</span> <span class="o">==</span> <span class="mi">6</span> <span class="ow">and</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
                <span class="n">ivl</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Interval</span><span class="p">(</span><span class="n">ivlstart</span><span class="p">,</span> <span class="n">ivlend</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ivls</span><span class="p">),</span> <span class="n">ivltype</span><span class="p">,</span> <span class="n">is_last_3exon_flag</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="p">[</span><span class="n">chromstrand</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ivl</span><span class="p">)</span>
                <span class="n">ivls</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ivl</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">set_range</span><span class="p">(</span><span class="n">genestart</span><span class="p">,</span> <span class="n">geneend</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">set_ivls</span><span class="p">(</span><span class="n">ivls</span><span class="p">)</span>
        <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">chromstrand</span><span class="p">,</span> <span class="n">ivls</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">ivls</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>  <span class="c1"># relies on the __lt__ method of vcy.Interval</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ivls</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">n</span></div>

<div class="viewcode-block" id="ExInCounter.mark_up_introns"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.mark_up_introns">[docs]</a>    <span class="k">def</span> <span class="nf">mark_up_introns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; Mark up introns that have reads across junctions to exons ?and count last exon reads? &quot;&quot;&quot;</span>
        <span class="c1"># Read the file</span>
        <span class="n">currchrom</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_unique_alignments</span><span class="p">(</span><span class="n">samfile</span><span class="p">):</span>
            <span class="c1"># Don&#39;t consider spliced reads (exonic) in this step</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">is_spliced</span><span class="p">():</span>
                <span class="k">continue</span>

            <span class="c1"># When the chromosome mapping of the read changes, change interval index.</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">chrom</span> <span class="o">!=</span> <span class="n">currchrom</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Marking up chromosome </span><span class="si">{r.chrom}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">currchrom</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">chrom</span>
                <span class="k">if</span> <span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;The .bam file refers to a chromosome &#39;</span><span class="si">{currchrom}</span><span class="s2">&#39; not present in the annotation (.gtf) file&quot;</span><span class="p">)</span>
                <span class="n">iif</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">IntervalsIndex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span><span class="p">])</span>
                <span class="n">iir</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">IntervalsIndex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span><span class="p">])</span>
            
            <span class="c1"># Consider the correct strand</span>
            <span class="n">ii</span> <span class="o">=</span> <span class="n">iif</span> <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">iir</span>

            <span class="c1"># Look for overlap between the intervals and the read</span>
            <span class="n">matchgenes</span><span class="p">,</span> <span class="n">matchivls</span> <span class="o">=</span> <span class="n">ii</span><span class="o">.</span><span class="n">find_overlapping_ivls</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matchgenes</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># No match, multiple-matches, chimera, or intron read accidentally spanning another genes&#39; exon zipped between the first genes&#39; exons</span>
                <span class="k">continue</span>

            <span class="n">matchgene</span> <span class="o">=</span> <span class="n">matchgenes</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="n">matchgene</span><span class="o">.</span><span class="n">add_read_stats</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>  <span class="c1"># NOTE: this might be removed if the stats output won&#39;t be used... but not sure</span>
            
            <span class="k">for</span> <span class="n">matchivl</span><span class="p">,</span> <span class="n">matchtype</span> <span class="ow">in</span> <span class="n">matchivls</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">matchtype</span> <span class="o">==</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_INSIDE</span><span class="p">:</span>
                    <span class="n">matchivl</span><span class="o">.</span><span class="n">add_read_inside</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">matchtype</span> <span class="o">&amp;</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_OVER5END</span><span class="p">:</span>  <span class="c1"># NOTE: should this be elif? or it makes sense to allow both?</span>
                    <span class="n">matchivl</span><span class="o">.</span><span class="n">add_read_spanning5end</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">matchtype</span> <span class="o">&amp;</span> <span class="n">vcy</span><span class="o">.</span><span class="n">MATCH_OVER3END</span><span class="p">:</span>  <span class="c1"># Added recently</span>
                    <span class="n">matchivl</span><span class="o">.</span><span class="n">add_read_spanning3end</span><span class="p">()</span>

        <span class="c1"># Validate the intron intervals on the basis of occurrences of Exon/Intron spanning reads</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">validate_intron_ivls</span><span class="p">()</span>  <span class="c1"># NOTE: this condition might be stricter than we want</span></div>

        <span class="c1"># NOTE: did we properly deal with the reads added to the genes so that self.count() does not get confused?</span>

<div class="viewcode-block" id="ExInCounter.count"><a class="viewcode-back" href="../../fullapi/api_cli_internals.html#velocyto.counter.ExInCounter.count">[docs]</a>    <span class="k">def</span> <span class="nf">count</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sam_output</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot; Do the counting of molecules&quot;&quot;&quot;</span>
        <span class="n">molitems</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># type: Dict[str, vcy.Molitem]  # keys are the cell_barcode + UMI, values are Molitem objects</span>
        <span class="n">currchrom</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">if</span> <span class="n">sam_output</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">f_sure_introns</span><span class="p">,</span> <span class="n">f_sure_exon</span><span class="p">,</span> <span class="n">f_maybe_exon</span><span class="p">,</span> <span class="n">f_others</span><span class="p">,</span> <span class="n">f_chimeras</span> <span class="o">=</span> <span class="n">sam_output</span>

        <span class="c1"># Loop through the aligment of the samfile</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_unique_alignments</span><span class="p">(</span><span class="n">samfile</span><span class="p">,</span> <span class="n">yield_line</span><span class="o">=</span><span class="n">sam_output</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">sam_output</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">r</span><span class="p">,</span> <span class="n">line</span> <span class="o">=</span> <span class="n">r</span>

            <span class="c1"># When the chromosome mapping of the read changes, change interval index.</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">chrom</span> <span class="o">!=</span> <span class="n">currchrom</span><span class="p">:</span>
                <span class="n">currchrom</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">chrom</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Counting on chr </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">currchrom</span><span class="p">)</span>
                <span class="n">iif</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">IntervalsIndex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span><span class="p">])</span>
                <span class="n">iir</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">IntervalsIndex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ivls_by_chromstrand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span><span class="p">])</span>
                <span class="n">iimaskf</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">IntervalsIndex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span><span class="p">])</span>
                <span class="n">iimaskr</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">IntervalsIndex</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="p">[</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span><span class="p">])</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;+&#39;</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span> <span class="ow">or</span> <span class="ow">not</span> <span class="p">(</span><span class="n">currchrom</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">mask_ivls_by_chromstrand</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{currchrom}</span><span class="s2"> is not repeat masked.&quot;</span><span class="p">)</span>

            <span class="c1"># Consider the correct strand</span>
            <span class="n">ii</span> <span class="o">=</span> <span class="n">iif</span> <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">iir</span>
            <span class="n">iim</span> <span class="o">=</span> <span class="n">iimaskf</span> <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">iimaskr</span>

            <span class="c1"># Check if read is fully inside a repeat, in that case skip it</span>
            <span class="k">if</span> <span class="n">iim</span><span class="o">.</span><span class="n">has_ivls_enclosing</span><span class="p">(</span><span class="n">r</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="c1"># Look for overlap between the intervals and the read</span>
            <span class="n">matchgenes</span><span class="p">,</span> <span class="n">matchivls</span> <span class="o">=</span> <span class="n">ii</span><span class="o">.</span><span class="n">find_overlapping_ivls</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
            <span class="c1"># logging.debug(&quot;matchgenes : %s&quot; % matchgenes)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matchgenes</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># No match, multiple-matches, chimera, or intron read accidentally spanning another genes&#39; exon zipped between the first genes&#39; exons</span>
                <span class="k">if</span> <span class="n">sam_output</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Chimera inside read or no matching gene&quot;</span><span class="p">)</span>
                    <span class="n">f_chimeras</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">continue</span>
            
            <span class="c1"># Count molecules by dealing with barcodes</span>
            <span class="n">bcumi</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">bc</span> <span class="o">+</span> <span class="n">r</span><span class="o">.</span><span class="n">umi</span>
            <span class="n">matchgene</span> <span class="o">=</span> <span class="n">matchgenes</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>  <span class="c1"># NOTE: how does it compare to matchgenes[0]? Is there some gc issue?</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Check if the read&#39;s gene is the same for the same molecule (bc+umi)</span>
                <span class="n">molitem</span> <span class="o">=</span> <span class="n">molitems</span><span class="p">[</span><span class="n">bcumi</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">molitem</span><span class="o">.</span><span class="n">gene</span> <span class="o">!=</span> <span class="n">matchgene</span><span class="p">:</span>
                    <span class="c1"># In that case ignore it</span>
                    <span class="k">if</span> <span class="n">sam_output</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Chimera between different reads of same molecule&quot;</span><span class="p">)</span>
                        <span class="n">f_chimeras</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                    <span class="k">continue</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="c1"># The bc/umi combination was seen for the first time, create a Molecule Item</span>
                <span class="n">molitem</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">Molitem</span><span class="p">(</span><span class="n">matchgene</span><span class="p">)</span>
                <span class="n">molitems</span><span class="p">[</span><span class="n">bcumi</span><span class="p">]</span> <span class="o">=</span> <span class="n">molitem</span>

            <span class="c1"># Add info about which interval was hit and wether it gives evidence of splicing</span>
            <span class="n">molitem</span><span class="o">.</span><span class="n">mark_hit_ivls</span><span class="p">(</span><span class="n">matchivls</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">r</span><span class="o">.</span><span class="n">is_spliced</span><span class="p">())</span>
            
            <span class="k">if</span> <span class="n">sam_output</span><span class="p">:</span>
                <span class="c1"># Resolve what the read is for the output sam files</span>
                <span class="c1"># NOTE: It would be nice to be able to output different sam files by molecule type other than read type</span>
                <span class="n">cum_is_maybe_exon</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">cum_is_sure_exon</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">cum_is_sure_intron</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">matchivls</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">cum_is_maybe_exon</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">is_maybe_exon</span><span class="p">)</span>
                    <span class="n">cum_is_sure_exon</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">is_sure_exon</span><span class="p">)</span>
                    <span class="n">cum_is_sure_intron</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">k</span><span class="o">.</span><span class="n">is_sure_intron</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">cum_is_sure_intron</span><span class="p">):</span>
                    <span class="n">f_sure_introns</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">cum_is_sure_exon</span><span class="p">):</span>
                    <span class="n">f_sure_exon</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">cum_is_maybe_exon</span><span class="p">):</span>
                    <span class="n">f_maybe_exon</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

        <span class="c1"># After the whole file has been read, do the actual counting</span>
        <span class="k">for</span> <span class="n">bcumi</span><span class="p">,</span> <span class="n">molitem</span> <span class="ow">in</span> <span class="n">molitems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">bc</span> <span class="o">=</span> <span class="n">bcumi</span><span class="p">[:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span><span class="p">]</span>  <span class="c1"># extract the bc part from the bc+umi</span>
            <span class="n">bcidx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_bcs2idx</span><span class="p">[</span><span class="n">bc</span><span class="p">]</span>
            <span class="n">molitem</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">bcidx</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Linnarsson Lab.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.9.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>