

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>velocyto.analysis module &mdash; velocyto 0.9.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="velocyto 0.9.5 documentation" href="../index.html"/>
        <link rel="up" title="API" href="index.html"/>
        <link rel="next" title="velocyto.estimation module" href="api_estimation.html"/>
        <link rel="prev" title="Command Line Tool API" href="cliapi.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> velocyto
          

          
          </a>

          
            
            
              <div class="version">
                0.9
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to velocyto.py!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install/index.html">Installation Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/index.html">Tutorial</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#command-line-api">Command Line API</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#library-api">Library API</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="index.html#the-analysis-module">The Analysis module</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">velocyto.analysis module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#the-estimation-module">The Estimation module</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#the-neighbors-module">The Neighbors module</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#the-diffusion-module">The Diffusion module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#internals">Internals</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../notebooks/index.html">Notebooks</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../citing/index.html">Authors and citations</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">velocyto</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">API</a> &raquo;</li>
        
      <li>velocyto.analysis module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/fullapi/api_analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-velocyto.analysis">
<span id="velocyto-analysis-module"></span><h1>velocyto.analysis module<a class="headerlink" href="#module-velocyto.analysis" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="velocyto.analysis.VelocytoLoom">
<em class="property">class </em><code class="descclassname">velocyto.analysis.</code><code class="descname">VelocytoLoom</code><span class="sig-paren">(</span><em>loom_filepath: str</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>A convenient object to store the data of a velocyto loom file.</p>
<p>Data will be stored in memory</p>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">velocyto</span> <span class="k">as</span> <span class="nn">vcy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vlm</span> <span class="o">=</span> <span class="n">vcy</span><span class="o">.</span><span class="n">VelocytoLoom</span><span class="p">(</span><span class="s2">&quot;foo.loom&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.S">
<code class="descname">S</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.S" title="Permalink to this definition">¶</a></dt>
<dd><p><em>np.ndarray</em> – Expressed spliced molecules</p>
</dd></dl>

<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.U">
<code class="descname">U</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.U" title="Permalink to this definition">¶</a></dt>
<dd><p><em>np.ndarray</em> – Unspliced molecule count</p>
</dd></dl>

<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.A">
<code class="descname">A</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.A" title="Permalink to this definition">¶</a></dt>
<dd><p><em>np.ndarray</em> – Anbiguous molecule count</p>
</dd></dl>

<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.ca">
<code class="descname">ca</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.ca" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict</em> – Column attributes of the loom file</p>
</dd></dl>

<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.ra">
<code class="descname">ra</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.ra" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict</em> – Row attributes of the loom file</p>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.adjust_totS_totU">
<code class="descname">adjust_totS_totU</code><span class="sig-paren">(</span><em>skip_low_U_pop: bool = True</em>, <em>normalize_total: bool = False</em>, <em>fit_with_low_U: bool = True</em>, <em>svr_C: float = 100</em>, <em>svr_gamma: float = 1e-06</em>, <em>plot: bool = True</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.adjust_totS_totU"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.adjust_totS_totU" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit a smooth line to the relation S_sz_tot and U_sz_tot</p>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.arrows_from_hidim_to_embedding">
<code class="descname">arrows_from_hidim_to_embedding</code><span class="sig-paren">(</span><em>hidim: str = 'pcsw'</em>, <em>ndims: int = 5</em>, <em>n_neighbors: int = 80</em>, <em>max_dist_hidim: float = None</em>, <em>embed: str = 'tsW'</em>, <em>max_dist_embed: float = None</em>, <em>renormalize_to_norm: bool = False</em>, <em>n_jobs: int = 4</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.arrows_from_hidim_to_embedding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.arrows_from_hidim_to_embedding" title="Permalink to this definition">¶</a></dt>
<dd><p>Project the arrows from an high dimensional space to a low dimensional space by kNN decomposition</p>
<p>DEPRECATED</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>hidim</strong> (<em>str</em>) – The name of the attribute containing the high dimensional space. It will be retrieved as getattr(self, hidim)
The updated vector at time t is assumed to be getattr(self, hidim + “_t”)</li>
<li><strong>ndims</strong> (<em>int</em>) – The number of dimensions of the high dimensional space to work in when decomposing and calculating kNN.</li>
<li><strong>n_neighbors</strong> (<em>int</em>) – The number of neighbours to take into account</li>
<li><strong>max_dist_hidim</strong> (<em>float</em><em>, </em><em>default=None</em>) – The maximum distance that is accepted for the kNN graph before dropping the neighbour
If None it will be set to 0.25 * average_distance_two_points_taken_at_random</li>
<li><strong>embed</strong> (<em>str</em>) – The name of the attribute containing the embedding. It will be retrieved as getattr(self, embed)</li>
<li><strong>max_dist_embed</strong> (<em>float</em><em>, </em><em>default=None</em>) – The maximum distance allowed
If None it will be set to 0.25 * average_distance_two_points_taken_at_random</li>
<li><strong>renormalize_to_norm</strong> (<em>bool</em><em>, </em><em>default=False</em>) – If true then the sum of the projections need to have the same magnitude as the orginal vector.
This forces the arrow to point within the manifiold so it is more conservative to set it to False</li>
<li><strong>n_jobs</strong> (<em>int</em>) – number of jobs to calkulate knn</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.calculate_embedding_shift">
<code class="descname">calculate_embedding_shift</code><span class="sig-paren">(</span><em>sigma_corr: float = 0.05</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.calculate_embedding_shift"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.calculate_embedding_shift" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.calculate_grid_arrows">
<code class="descname">calculate_grid_arrows</code><span class="sig-paren">(</span><em>embed: str = 'embedding'</em>, <em>smooth: float = 0.5</em>, <em>steps: typing.Tuple = (40</em>, <em>40)</em>, <em>n_neighbors: int = 100</em>, <em>n_jobs: int = 4</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.calculate_grid_arrows"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.calculate_grid_arrows" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the velocity using a points on a regular grid and a gaussian kernel</p>
<p>Note: the function should work also for n-dimensional grid</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>embed</strong> (<em>str</em><em>, </em><em>default=embedding</em>) – The name of the attribute containing the embedding. It will be retrieved as getattr(self, embed)
The updated vector at time t is assumed to be getattr(self, embed + ‘_t’)
Or the difference vector is getattr(self, ‘delta’ + ‘_’ + embed)</li>
<li><strong>smooth</strong> (<em>float</em><em>, </em><em>smooth=0.5</em>) – Higher value correspond to thaking in consideration further points
the standardad deviation of the gaussian karnel is smooth * stepsize</li>
<li><strong>steps</strong> (<em>tuple</em><em>, </em><em>detaul</em>) – the number of steps in the grid for each axis</li>
<li><strong>n_neighbors</strong> – number of neighbours to use in the calculation, bigger number should not change too much the results..
…as soon as smooth is small
Higher value correspond to slower exectution time</li>
<li><strong>n_jobs</strong> – number of processes for parallel computing</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><em>Nothing but it sets the attributes</em></li>
<li><em>flow_embedding, flow_grid, flow, flow_magnitude</em></li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.calculate_local_correlations">
<code class="descname">calculate_local_correlations</code><span class="sig-paren">(</span><em>hidim: str = 'pcsw'</em>, <em>embed: str = 'tsW'</em>, <em>transform: str = None</em>, <em>ndims: int = None</em>, <em>n_neighbors: int = 80</em>, <em>psc: float = 1.0</em>, <em>knn_random: bool = False</em>, <em>sampled_fraction: float = 0.1</em>, <em>delta_kind: str = 'clipped'</em>, <em>sampling_pobs: typing.Tuple[float</em>, <em>float] = (0.5</em>, <em>0.1)</em>, <em>max_dist_embed: float = None</em>, <em>n_jobs: int = 4</em>, <em>random_seed: int = 15071990</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.calculate_local_correlations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.calculate_local_correlations" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses correlation to estimate transition probabilities for every cells to its embedding neighborhood</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>hidim</strong> (<em>str</em>) – The name of the attribute containing the high dimensional space. It will be retrieved as getattr(self, hidim)
The updated vector at time t is assumed to be getattr(self, hidim + “_t”)
Appending .T to the string will transpose the matrix (usefull in case we want to use S or Sx)</li>
<li><strong>embed</strong> (<em>str</em>) – The name of the attribute containing the embedding. It will be retrieved as getattr(self, embed)</li>
<li><strong>transform</strong> (<em>str</em><em>, </em><em>default=None</em>) – The transformation that is applies on the high dimensional space.
If None the raw data will be used</li>
<li><strong>ndims</strong> (<em>int</em><em>, </em><em>default=None</em>) – The number of dimensions of the high dimensional space to work with. If None all will be considered
It makes sense only when using principal components</li>
<li><strong>n_neighbors</strong> (<em>int</em><em>, </em><em>default=80</em>) – The number of neighbours to take into account</li>
<li><strong>delta_kind</strong> (<em>clipped</em><em>, </em><em>unclipped</em><em>, </em><em>residual</em>) – </li>
<li><strong>max_dist_embed</strong> (<em>float</em><em>, </em><em>default=None</em>) – CURRENTLY NOT USED
The maximum distance allowed
If None it will be set to 0.25 * average_distance_two_points_taken_at_random</li>
<li><strong>n_jobs</strong> (<em>int</em><em>, </em><em>default=4</em>) – number of jobs to calkulate knn</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.calculate_shift">
<code class="descname">calculate_shift</code><span class="sig-paren">(</span><em>kind: str = 'log2_ratio'</em>, <em>assumption: str = 'constant_velocity'</em>, <em>delta_t: float = 1</em>, <em>min_S: float = 0.0001</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.calculate_shift"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.calculate_shift" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the change in gene expression for every cell</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>kind</strong> (<em>str</em><em>, </em><em>default=&quot;log2_ratio&quot;</em>) – difference
log2_ratio
log2_ratioW</li>
<li><strong>assumption</strong> (<em>str</em><em>, </em><em>default=&quot;constant_velocity&quot;</em>) – constant_velocity
constant_unspliced</li>
<li><strong>min_S</strong> (<em>float</em><em>, </em><em>default=min_S</em>) – constant small value added in case of log2_ratio estimation</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">.delta_S</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Nothing it only creates the following attributes</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.calculate_velocity">
<code class="descname">calculate_velocity</code><span class="sig-paren">(</span><em>kind: str = 'M-value'</em>, <em>pcount: float = 1</em>, <em>size_normalized: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.calculate_velocity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.calculate_velocity" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate velocity or related parameter parameter</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>kind</strong> (<em>str</em><em>, </em><em>default=&quot;M-value&quot;</em>) – “M-value” calculates the M value as log2(U_measured) - log2(U_predicted)
“M-valueKNN” calculates the M value as using gene KNN
“residual” calculates the velocity as U_measured) - U_predicted
“MW-value” calculates the M value as log2(WUpred.dot(H)) - np.log2(WU.dot(H))</li>
<li><strong>pcount</strong> (<em>float</em><em>, </em><em>default=1</em>) – (only used in kind=”M-value”) pseudocount to add to the reads/molecule counts before log2 transforming</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.cluster_ix">
<code class="descname">cluster_ix</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.cluster_ix" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="velocyto.analysis.VelocytoLoom.cluster_uid">
<code class="descname">cluster_uid</code><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.cluster_uid" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.custom_filter_genes">
<code class="descname">custom_filter_genes</code><span class="sig-paren">(</span><em>attr_names: typing.List[str], bool_filter: numpy.ndarray, keep_unfiltered: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.custom_filter_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.custom_filter_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>It filter genes from multiple attributes, taking care if they are dictionaries or numpy arrays
and if the unfiltered version need to be mantained</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>attr_names</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – a list of the attributes to be modified. The can be
1d arrays
dictionary of 1d arrays
nd arrays, will be filtered by axis=0
if .T is specified by axis=-1</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.extrapolate_cell_at_t">
<code class="descname">extrapolate_cell_at_t</code><span class="sig-paren">(</span><em>delta_t: float = 0.5</em>, <em>clip: bool = True</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.extrapolate_cell_at_t"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.extrapolate_cell_at_t" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the value for the cell after time delta_t</p>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.filter_cells">
<code class="descname">filter_cells</code><span class="sig-paren">(</span><em>bool_array: numpy.ndarray</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.filter_cells"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.filter_cells" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter cells that will be kept in the analysis</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>bool_array</strong> (<em>np.ndarray</em><em> (</em><em>size</em><em> )</em>) – array describing the cells to keep (True)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">Nothing but it remove some cells from S and U</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.filter_genes">
<code class="descname">filter_genes</code><span class="sig-paren">(</span><em>by_detection_levels: bool = False</em>, <em>by_cluster_expression: bool = False</em>, <em>by_cv_vs_mean: bool = False</em>, <em>by_custom_array: typing.Any = None</em>, <em>keep_unfiltered: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.filter_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.filter_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter genes taking care that all the matrixes and all the connected annotation get filtered accordingly</p>
<p>Attributes affected: .U, .S, .ra
if present also: .ei_stats_matrix, .ei_stats_ra</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>by_detection_levels</strong> (<em>bool</em><em>, </em><em>default=False</em>) – filter genes by the score_detection_levels result</li>
<li><strong>by_cluster_expression</strong> (<em>bool</em><em>, </em><em>default=True</em><em> (</em><em>NOTE: in legacy code was True</em><em>)</em>) – filter genes by the score_cluster_expression result</li>
<li><strong>by_cv_vs_mean</strong> (<em>bool</em><em>, </em><em>default=True</em><em> (</em><em>NOTE: in legacy code was True</em><em>)</em>) – filter genes by the score_cluster_expression result</li>
<li><strong>np.ndarray</strong><strong>, </strong><strong>default=None</strong> (<em>by_custom_array</em><em>,</em>) – provide a booelan or index array</li>
<li><strong>keep_unfiltered</strong> (<em>bool</em><em>, </em><em>default=False</em>) – wether to create attributes self.S_prefilter, self.U_prefilter, self.ra_prefilter,
optionally if present (self.ei_stats_matrix_prefilter, ei_stats_ra_prefilter)
(array will be made sparse to minimize memory footprint)
or just overwrite the previous arrays</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><em>Nothing but it updates the self.S, self.U, self.ra attributes</em></li>
<li><em>(if present also self.ei_stats_matrix, ei_stats_ra)</em></li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.filter_genes_good_fit">
<code class="descname">filter_genes_good_fit</code><span class="sig-paren">(</span><em>minR: float = 0.1</em>, <em>min_gamma: float = 0.01</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.filter_genes_good_fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.filter_genes_good_fit" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.fit_gammas">
<code class="descname">fit_gammas</code><span class="sig-paren">(</span><em>steady_state_bool: numpy.ndarray = None, use_imputed_data: bool = True, use_size_norm: bool = False, single_q: bool = True, fixperc_q: bool = False, weighted: bool = False, weights: numpy.ndarray = 'maxmin_diag', limit_gamma: bool = False, maxmin_perc: typing.List[float] = [2, 98], maxmin_weighted_pow: float = 15</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.fit_gammas"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.fit_gammas" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit gamma without intercepts of the gene components</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>q_positive</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Wether to constrain q to positive values</li>
<li><strong>multiple_qs</strong> – Fit with one offset for each neighborhood</li>
<li><strong>single_q</strong> – Fit with offset</li>
<li><strong>weighted</strong> – weighted fit without offset</li>
<li><strong>weights</strong> – values that get multiplied to the squared residuals</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.gene_knn_imputation">
<code class="descname">gene_knn_imputation</code><span class="sig-paren">(</span><em>metric: str = 'correlation'</em>, <em>k: int = 15</em>, <em>diag: float = 1</em>, <em>scale_weights: bool = True</em>, <em>balanced: bool = True</em>, <em>b_sight: int = 100</em>, <em>b_maxl: int = 18</em>, <em>pca_space: float = False</em>, <em>n_jobs: int = 8</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.gene_knn_imputation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.gene_knn_imputation" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs genes k-nn imputation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>metric</strong> (<em>str</em><em>, </em><em>default=&quot;correlation&quot;</em>) – “euclidean” or “correlation”</li>
<li><strong>k</strong> (<em>int</em><em>, </em><em>defaut=15</em>) – number of neighbours</li>
<li><strong>diag</strong> (<em>int</em><em>, </em><em>default=1</em>) – before smoothing this value is substituted in the diagonal of the knn contiguity matrix
Resulting in a reduction of the smoothing effect
E.g. if diag=8 and k=10 value of Si = (8 * S_i + sum(S_n, with n in 5nn of i)) / (8+5)</li>
<li><strong>scale_weights</strong> (<em>bool</em><em>, </em><em>default=True</em>) – wether to scale weights by gene total expression/yield</li>
<li><strong>balanced</strong> (<em>bool</em><em>, </em><em>dafault=True</em>) – wheter to use BalancedKNN verison</li>
<li><strong>b_sight</strong> (<em>int</em><em>, </em><em>default=100</em>) – the sight parameter of BalancedKNN (used only if balanced == True)</li>
<li><strong>b_maxl</strong> (<em>int</em><em>, </em><em>default=18</em>) – the maxl patamenter of BalancedKNN (used only if balanced == True)</li>
<li><strong>pca_space</strong> (<em>bool</em><em>, </em><em>default=False</em>) – if True the knn will be performed on self.pca
otherwise it will use self.S_norm (log2 size normalized data)</li>
<li><strong>n_jobs</strong> (<em>int</em><em>, </em><em>default=8</em>) – number of parallel jobs in knn calculation</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><em>Nothing but it creates the attributes</em></li>
<li><strong>`gknn`</strong> (<em>sparse matrix</em>)</li>
<li><strong>`gknn_smoothig_w`</strong> (<em>the weights used for the smoothing</em>)</li>
<li><strong>`Sx`, `Ux`</strong> (<em>the imputed values</em>)</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.knn_imputation">
<code class="descname">knn_imputation</code><span class="sig-paren">(</span><em>pca_space: float = True</em>, <em>diag: float = 5</em>, <em>k: int = 5</em>, <em>n_pca_dims: int = None</em>, <em>maximum: bool = False</em>, <em>metric: str = 'euclidean'</em>, <em>size_norm: bool = True</em>, <em>balanced: bool = False</em>, <em>b_sight: int = 100</em>, <em>b_maxl: int = 7</em>, <em>n_jobs: int = 8</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.knn_imputation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.knn_imputation" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs k-nn imputation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pca_space</strong> (<em>bool</em><em>, </em><em>default=True</em>) – if True the knn will be performed on self.pca
otherwise it will use self.S_norm (log2 size normalized data)</li>
<li><strong>diag</strong> (<em>int</em>) – before smoothing this value is substituted in the diagonal of the knn contiguity matrix
Resulting in a reduction of the smoothing effect
E.g. if diag=8 and k=10 value of Si = (8 * S_i + sum(S_n, with n in 5nn of i)) / (8+5)</li>
<li><strong>maximum</strong> (<em>bool</em>) – wether to take the maximum value of the smoothing and the original matrix</li>
<li><strong>metric</strong> (<em>str</em>) – “euclidean” or “correlation”</li>
<li><strong>k</strong> (<em>int</em>) – number of neighbours</li>
<li><strong>n_pca_dims</strong> (<em>int</em><em>, </em><em>default=None</em>) – number of pca to use for the knn distance metric. If None all pcs will be used. (used only if pca_space == True)</li>
<li><strong>balanced</strong> (<em>bool</em>) – wheter to use BalancedKNN verison</li>
<li><strong>b_sight</strong> (<em>int</em>) – the sight parameter of BalancedKNN (used only if balanced == True)</li>
<li><strong>b_maxl</strong> (<em>int</em>) – the maxl patamenter of BalancedKNN (used only if balanced == True)</li>
<li><strong>n_jobs</strong> (<em>int</em><em>, </em><em>default 8</em>) – number of parallel jobs in knn calculation</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><em>Nothing but it creates the attributes</em></li>
<li><strong>`knn`</strong> (<em>sparse matrix</em>)</li>
<li><strong>`knn_smoothig_w`</strong> (<em>the weights used for the smoothing</em>)</li>
<li><strong>`Sx`, `Ux`</strong> (<em>the imputed values</em>)</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.knn_imputation_precomputed">
<code class="descname">knn_imputation_precomputed</code><span class="sig-paren">(</span><em>knn_smoothig_w: scipy.sparse.lil.lil_matrix</em>, <em>maximum: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.knn_imputation_precomputed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.knn_imputation_precomputed" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs k-nn imputation (like <cite>.knn_imputation()</cite>) but with a precomputed weight matrix</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>knn_smoothig_w</strong> (<em>sparse.lil_matrix</em>) – the sparse matrix to be convolved with self.S_sz and self.U_sz
This should be the result of somethign like:
connectivity.setdiag(diagonal_value)
knn_smoothig_w = connectivity_to_weights(connectivity)</li>
<li><strong>maximum</strong> (<em>bool</em><em>, </em><em>default=False</em>) – wether to take the maximum value of the smoothing and the original matrix</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><em>Nothing but it creates the attributes</em></li>
<li><strong>Sx, Ux</strong> (<em>the imputed values</em>)</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.normalize">
<code class="descname">normalize</code><span class="sig-paren">(</span><em>which: str = 'both'</em>, <em>size: bool = True</em>, <em>log: bool = True</em>, <em>pcount: float = 1</em>, <em>relative_size: numpy.ndarray = None</em>, <em>use_S_size_for_U: bool = False</em>, <em>target_size: typing.Tuple[float</em>, <em>float] = (None</em>, <em>None)</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.normalize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.normalize" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalization for the count data</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>which</strong> (<em>either 'both'</em><em>, </em><em>'S'</em><em>, </em><em>'U'</em><em>, </em><em>&quot;imputed&quot;</em><em>, </em><em>&quot;Sx&quot;</em><em>, </em><em>&quot;Ux&quot;</em>) – which attributes to normalize.
“both” corresponds to “S” and “U”
“imputed” corresponds to “Sx” and “Ux”</li>
<li><strong>size</strong> (<em>bool</em>) – perform size normalization</li>
<li><strong>log</strong> (<em>bool</em>) – perform log normalization (if size==True, this comes after the size normalization)</li>
<li><strong>pcount</strong> (<em>int</em><em>, </em><em>default: 1</em>) – The extra count added when logging (log2)</li>
<li><strong>relative_size</strong> (<em>np.ndarray</em><em>, </em><em>default=None</em>) – if None it calculate the sums the mulecules per cell (self.S.sum(0))
if an array is provided it use it for the normalization</li>
<li><strong>use_S_size_for_U</strong> (<em>bool</em>) – U is size normalized using the sum of molecules of S</li>
<li><strong>target_size</strong> (<em>float</em><em> or </em><em>Tuple</em><em>[</em><em>float</em><em>, </em><em>float</em><em>] </em><em>(</em><em>depending if the which parameter implies 1</em><em> or </em><em>more normalizations</em><em>)</em>) – the size of the cells after normalization will be set to.
If tuple the order is (S, U) or (Sx, Ux)
If None the target size is the average of the cell sizes</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li>Nothing but creates the attributes <cite>U_norm</cite>, <cite>U_sz</cite> and <cite>S_norm</cite>, “S_sz”</li>
<li>or <cite>Ux_norm</cite>, <cite>Ux_sz</cite> and <cite>Sx_norm</cite>, “Sx_sz”</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.normalize_by_total">
<code class="descname">normalize_by_total</code><span class="sig-paren">(</span><em>min_perc_U: float = 0.5</em>, <em>plot: bool = True</em>, <em>skip_low_U_pop: bool = True</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.normalize_by_total"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.normalize_by_total" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize the cells using the (initial) total moluecules as size estimate</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>min_perc_U</strong> (<em>float</em>) – the percentile to use as a minimum value allowed for the size normalization</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.normalize_median">
<code class="descname">normalize_median</code><span class="sig-paren">(</span><em>which: str = 'imputed'</em>, <em>skip_low_U_pop: bool = True</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.normalize_median"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.normalize_median" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.perform_PCA">
<code class="descname">perform_PCA</code><span class="sig-paren">(</span><em>which: str = 'S_norm'</em>, <em>n_components: int = None</em>, <em>div_by_std: bool = False</em>, <em>mult_by_scalefact: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.perform_PCA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.perform_PCA" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.perform_PCA_imputed">
<code class="descname">perform_PCA_imputed</code><span class="sig-paren">(</span><em>n_components: int = None</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.perform_PCA_imputed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.perform_PCA_imputed" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.perform_TSNE">
<code class="descname">perform_TSNE</code><span class="sig-paren">(</span><em>n_dims: int = 2</em>, <em>perplexity: float = 30</em>, <em>initial_pos: numpy.ndarray = None</em>, <em>theta: float = 0.5</em>, <em>n_pca_dim: int = None</em>, <em>max_iter: int = 1000</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.perform_TSNE"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.perform_TSNE" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform TSNE on the PCA using barnes hut approximation (C implementation)</p>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_arrows_embedding">
<code class="descname">plot_arrows_embedding</code><span class="sig-paren">(</span><em>choice: int = 1000</em>, <em>plot_scatter: bool = False</em>, <em>color_arrow: str = 'cluster'</em>, <em>quiver_scale: float = None</em>, <em>new_fig: bool = False</em>, <em>**quiver_kwargs: typing.Any</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_arrows_embedding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_arrows_embedding" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the results of the arrows_from_hidim_to_embedding</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>choice</strong> (<em>int</em><em>, </em><em>default = 1000</em>) – the number of cells to randomly pick to plot the arrows</li>
<li><strong>plot_scatter</strong> (<em>bool</em><em>, </em><em>default = False</em>) – wether to plot the points</li>
<li><strong>color_arrow</strong> (<em>str</em><em>, </em><em>default = &quot;cluster</em>) – the color of the arrows, if “cluster” the arrows are colored the same as the cluster</li>
<li><strong>epsilon</strong> (<em>float</em><em>, </em><em>default = None</em>) – the minimal size of the arrow that can be plotted before it becomes invisible</li>
<li><strong>quiver_scale</strong> (<em>float</em>) – scale argument passed to the matplotlib quiver function</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Nothing, just plots the tsne with arrows</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_cell_transitions">
<code class="descname">plot_cell_transitions</code><span class="sig-paren">(</span><em>cell_ix: int = 0</em>, <em>alpha: float = 0.1</em>, <em>alpha_neigh: float = 0.2</em>, <em>cmap_name: str = 'RdBu_r'</em>, <em>plot_arrow: bool = True</em>, <em>mark_cell: bool = True</em>, <em>head_width: int = 3</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_cell_transitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_cell_transitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the probability of a cell to transition to any other cell</p>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_expression_as_color">
<code class="descname">plot_expression_as_color</code><span class="sig-paren">(</span><em>gene_name: str = None</em>, <em>imputed: bool = True</em>, <em>cmap: typing.Any = &lt;matplotlib.colors.LinearSegmentedColormap object&gt;</em>, <em>gs: typing.Any = None</em>, <em>which_tsne: str = 'ts'</em>, <em>**kwargs: typing.Dict</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_expression_as_color"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_expression_as_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot expression (imputed) as color on the Tsne</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gene_name</strong> (<em>str</em>) – The name of the gene, should be present in self.S</li>
<li><strong>cmap</strong> (<em>maplotlib.cm.Colormap</em>) – Colormap to use, devergent are better, RdBu_r is default
Notice that 0 will be always set as the center of the colormap. (e.g. white in RdBu_r)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Nothing</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_fractions">
<code class="descname">plot_fractions</code><span class="sig-paren">(</span><em>save2file: str = None</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_fractions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_fractions" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots a barplot of the abundance of spliced/unspliced of molecules in the dataset</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>save2file</strong> (<em>str</em><em> (</em><em>default: None</em><em>)</em>) – If not None specifies the file path to wich plots get saved</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_grid_arrows">
<code class="descname">plot_grid_arrows</code><span class="sig-paren">(</span><em>quiver_scale: float = None</em>, <em>min_mass: float = 1</em>, <em>min_magnitude: float = None</em>, <em>scatter_kwargs_dict: typing.Dict = None</em>, <em>plot_dots: bool = False</em>, <em>**quiver_kwargs: typing.Any</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_grid_arrows"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_grid_arrows" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_pca">
<code class="descname">plot_pca</code><span class="sig-paren">(</span><em>dim: typing.List[int] = [0, 1, 2], elev: float = 60, azim: float = -140</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_pca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_pca" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_pca_imputed">
<code class="descname">plot_pca_imputed</code><span class="sig-paren">(</span><em>dim: typing.List[int] = [0, 1, 2], elev: float = 60, azim: float = -140</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_pca_imputed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_pca_imputed" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.plot_velocity_as_color">
<code class="descname">plot_velocity_as_color</code><span class="sig-paren">(</span><em>gene_name: str = None</em>, <em>cmap: typing.Any = &lt;matplotlib.colors.LinearSegmentedColormap object&gt;</em>, <em>gs: typing.Any = None</em>, <em>which_tsne: str = 'ts'</em>, <em>**kwargs: typing.Dict</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.plot_velocity_as_color"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.plot_velocity_as_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot velocity as color on the Tsne</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>which_tsne</strong> (<em>str</em>) – can be “tsW” or other tsne attribute accessible by getattr(self, which_tsne)</li>
<li><strong>cmap</strong> (<em>maplotlib.cm.Colormap</em>) – Colormap to use, devergent are better, RdBu_r is default
Notice that 0 will be always set as the center of the colormap. (e.g. white in RdBu_r)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Nothing</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.predict_U">
<code class="descname">predict_U</code><span class="sig-paren">(</span><em>which_gamma: str = 'gammas'</em>, <em>which_S: str = 'Sx_sz'</em>, <em>which_offset: str = None</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.predict_U"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.predict_U" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict U given the gamma model fit</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>which_gamma</strong> (<em>str</em><em>, </em><em>default=&quot;gammas&quot;</em>) – name of the attribute to use as gamma</li>
<li><strong>which_S</strong> (<em>str</em><em>, </em><em>default=&quot;Sx_sz&quot;</em>) – name of the attribute to use as S</li>
<li><strong>which_offset</strong> (<em>str</em><em>, </em><em>default=None</em>) – (optional) offset has to be used specify which attribute to use</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">.Upred</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Noting but it creates the attribute</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.reload_raw">
<code class="descname">reload_raw</code><span class="sig-paren">(</span><em>substitute: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.reload_raw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.reload_raw" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.score_cluster_expression">
<code class="descname">score_cluster_expression</code><span class="sig-paren">(</span><em>min_avg_U: float = 0.02</em>, <em>min_avg_S: float = 0.08</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.score_cluster_expression"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.score_cluster_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare filtering genes on the basis of knowledge on the cluster</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>min_avg_U</strong> (<em>float</em>) – Include genes that have unspliced average bigger than <cite>min_avg_U</cite> in at least one of the clusters</li>
<li><strong>min_avg_U</strong> – Include genes that have spliced average bigger than <cite>min_avg_U</cite> in at least one of the clusters</li>
<li><strong>Note</strong> (<em>the two conditions are combined by and &quot;&amp;&quot; logical operator</em>) – </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><strong>clu_avg_selected</strong> (<em>np.ndarray bool</em>) – The gene cluster that is selected</li>
<li><strong>Note</strong> (<em>an attribute with the same name is created</em>)</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.score_cv_vs_mean">
<code class="descname">score_cv_vs_mean</code><span class="sig-paren">(</span><em>N: int = 3000</em>, <em>min_expr_cells: int = 2</em>, <em>max_expr_avg: float = 20</em>, <em>svr_gamma: float = None</em>, <em>winsorize: bool = False</em>, <em>winsor_perc: typing.Tuple[float</em>, <em>float] = (1</em>, <em>99.5)</em>, <em>plot: bool = False</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.score_cv_vs_mean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.score_cv_vs_mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Rank genes on the basis of a CV vs mean fit, it uses a nonparametric fit (Support Vector Regression)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>N</strong> (<em>int</em>) – the number to select</li>
<li><strong>min_expr_cells</strong> (<em>int</em>) – minimum number of cells that express that gene for it to be considered in the fit</li>
<li><strong>max_expr_avg</strong> (<em>float</em>) – The maximum average accepted before discarding from the the gene as house-keeping/outlier</li>
<li><strong>svr_gamma</strong> (<em>float</em>) – the gamma hyperparameter of the SVR</li>
<li><strong>plot</strong> (<em>bool</em>) – wether to show a plot</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><strong>score</strong> (<em>np.ndarray</em>) – How much the observed CV is higher than the one predicted by a noise model fit to the data
# the same array will be stored as the attribure <cite>cv_mean_score</cite></li>
<li><strong>selected</strong> (<em>np.ndarray bool</em>) – on the basis of the N parameter
# the same array will be stored as the attribure <cite>cv_mean_selected</cite></li>
<li><strong>Note</strong> (<em>genes excluded from the fit will have in the output the same score as the less noisy gene in the dataset.</em>)</li>
<li><strong>NOTE</strong> (<em>The function returns more genes than N in case more genes with the same score as present. This may result</em>) – in a cv vs mean plot that looks weirdly colored red/blue when looked at low magnification</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.score_detection_levels">
<code class="descname">score_detection_levels</code><span class="sig-paren">(</span><em>min_expr_counts: int = 50</em>, <em>min_cells_express: int = 20</em>, <em>min_expr_counts_U: int = 0</em>, <em>min_cells_express_U: int = 0</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.score_detection_levels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.score_detection_levels" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare basic filtering of genes on the basis of their detection levels</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>min_expr_counts</strong> (<em>float</em>) – The minimum number of spliced molecules detected considering all the cells</li>
<li><strong>min_cells_express</strong> (<em>float</em>) – The minimum number of cells that express spliced molecules of a gene</li>
<li><strong>min_expr_counts_U</strong> (<em>float</em>) – The minimum number of unspliced molecules detected considering all the cells</li>
<li><strong>min_cells_express_U</strong> (<em>float</em>) – The minimum number of cells that express unspliced molecules of a gene</li>
<li><strong>Note</strong> (<em>the conditions are combined by and &quot;&amp;&quot; logical operator</em>) – </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">Nothing but an attribute self.detection_level_selected is created</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="velocyto.analysis.VelocytoLoom.set_clusters">
<code class="descname">set_clusters</code><span class="sig-paren">(</span><em>cluster_labels: numpy.ndarray</em>, <em>cluter_colors_dict: typing.Dict[typing.Any</em>, <em>typing.List[float]] = None</em>, <em>colormap: typing.Any = None</em><span class="sig-paren">)</span> &#x2192; None<a class="reference internal" href="../_modules/velocyto/analysis.html#VelocytoLoom.set_clusters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.VelocytoLoom.set_clusters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="velocyto.analysis.colormap_fun">
<code class="descclassname">velocyto.analysis.</code><code class="descname">colormap_fun</code><span class="sig-paren">(</span><em>x: numpy.ndarray</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="reference internal" href="../_modules/velocyto/analysis.html#colormap_fun"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.colormap_fun" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="velocyto.analysis.ixs_thatsort_a2b">
<code class="descclassname">velocyto.analysis.</code><code class="descname">ixs_thatsort_a2b</code><span class="sig-paren">(</span><em>a: numpy.ndarray</em>, <em>b: numpy.ndarray</em>, <em>check_content: bool = True</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="reference internal" href="../_modules/velocyto/analysis.html#ixs_thatsort_a2b"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.ixs_thatsort_a2b" title="Permalink to this definition">¶</a></dt>
<dd><p>This is super duper magic sauce to make the order of one list to be like another</p>
</dd></dl>

<dl class="function">
<dt id="velocyto.analysis.scale_to_match_median">
<code class="descclassname">velocyto.analysis.</code><code class="descname">scale_to_match_median</code><span class="sig-paren">(</span><em>sparse_matrix: scipy.sparse.csr.csr_matrix</em>, <em>genes_total: numpy.ndarray</em><span class="sig-paren">)</span> &#x2192; scipy.sparse.csr.csr_matrix<a class="reference internal" href="../_modules/velocyto/analysis.html#scale_to_match_median"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.scale_to_match_median" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalize contribution of different neighbor genes to match the median totals</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sparse_matrix</strong> (<em>sparse.csr_matrix</em>) – weights matrix</li>
<li><strong>genes_total</strong> (<em>sparse.csr_matrix shape=</em><em>(</em><em>sparse_matrix.shape</em><em>[</em><em>0</em><em>]</em><em>)</em>) – array of the total molecules detected for each gene</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><strong>knn_weights</strong> (<em>sparse.csr_matrix</em>) – sparse_matrix after the normalization</li>
<li><em># NOTE, since the use I made of this later I could have changed sparse_matrix in place</em></li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="velocyto.analysis.scatter_viz">
<code class="descclassname">velocyto.analysis.</code><code class="descname">scatter_viz</code><span class="sig-paren">(</span><em>x: numpy.ndarray</em>, <em>y: numpy.ndarray</em>, <em>*args: typing.Any</em>, <em>**kwargs: typing.Any</em><span class="sig-paren">)</span> &#x2192; typing.Any<a class="reference internal" href="../_modules/velocyto/analysis.html#scatter_viz"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#velocyto.analysis.scatter_viz" title="Permalink to this definition">¶</a></dt>
<dd><p>A wrapper of scatter plot that guarantees that every point is visible in a very crowded scatterplot</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>x</strong> (<em>np.ndarray</em>) – x axis coordiantes</li>
<li><strong>y</strong> (<em>np.ndarray</em>) – y axis coordiantes</li>
<li><strong>and kwargs</strong> (<em>args</em>) – positional and keyword arguments as in matplotplib.pyplot.scatter</li>
<li><strong>Retruns</strong> – </li>
<li><strong>-------</strong> – </li>
<li><strong>the graph and returns the axes object</strong> (<em>Plots</em>) – </li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="api_estimation.html" class="btn btn-neutral float-right" title="velocyto.estimation module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="cliapi.html" class="btn btn-neutral" title="Command Line Tool API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Linnarsson Lab.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'0.9.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>